### PR TITLE
added initial transfer subsystem

### PR DESCRIPTION
**Pull request by yutytuty**

### CODE REVIEW

File: src/main/java/frc/robot/RobotMap.java [hunk 1]
```diff
@@ -1,4 +1,12 @@
 package frc.robot;
 
 public class RobotMap {
+    public interface CAN {
+        int TRANSFER_TIMING_BELT_TALON = -1;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/RobotMap.java:5]: just `TRANSFER_TALON` is enough

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 2]
```diff
@@ -1,4 +1,44 @@
 package frc.robot.Subsystems;
 
-public class Transfer {
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.spikes2212.command.genericsubsystem.GenericSubsystem;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.RobotMap;
+
+public class Transfer extends GenericSubsystem {
+    WPI_TalonSRX talon;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:9]: private?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 3]
```diff
@@ -1,4 +1,44 @@
...
+    DigitalInput entranceSensor;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:10]: private?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 4]
```diff
@@ -1,4 +1,44 @@
...
+    DigitalInput exitSensor;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:11]: private?

File: src/main/java/frc/robot/Robot.java [hunk 5]
```diff
@@ -19,10 +20,10 @@
      * This function is run when the robot is first started up and should be used for any
      * initialization code.
      */
+    Transfer transfer;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:23]: private?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 6]
```diff
@@ -1,4 +1,44 @@
...
+
+    private static Transfer instance;
+
+    private Transfer() {
+        super();
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:16]: why?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 7]
```diff
@@ -1,4 +1,44 @@
...
+        this.talon = new WPI_TalonSRX(RobotMap.CAN.TRANSFER_TIMING_BELT_TALON);
+        this.entranceSensor = new DigitalInput(RobotMap.DIO.TRANSFER_ENTRANCE_LIGHT_SENSOR);
+        this.exitSensor = new DigitalInput(RobotMap.DIO.TRANSFER_EXIT_LIMIT_SWITCH);
+    }
+
+    public static Transfer getInstance() {
+        if (instance == null) {
+            instance = new Transfer();
+        }
+
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:26]: why blank line?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 8]
```diff
@@ -1,4 +1,44 @@
 package frc.robot.Subsystems;
 
-public class Transfer {
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.spikes2212.command.genericsubsystem.GenericSubsystem;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.RobotMap;
+
+public class Transfer extends GenericSubsystem {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:8]: change to `MotoredGenericSubsystem`

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 9]
```diff
@@ -1,4 +1,44 @@
...
+
+public class Transfer extends MotoredGenericSubsystem {
+    private WPI_TalonSRX talon;
+    private DigitalInput entranceSensor;
+    private DigitalInput exitSensor;
+
+    private static Transfer instance;
+
+    private Transfer(WPI_TalonSRX talon) {
+        super("Transfer", talon);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:17]: lowercase T in the `Transform`

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 10]
```diff
@@ -1,4 +1,44 @@
...
+
+    public static Transfer getInstance() {
+        if (instance == null) {
+            instance = new Transfer(new WPI_TalonSRX(RobotMap.CAN.TRANSFER_TALON));
+        }
+        return instance;
+    }
+
+    @Override
+    protected void apply(double speed) {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:31]: why apply?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 11]
```diff
@@ -1,4 +1,44 @@
...
+        talon.set(speed);
+    }
+
+    @Override
+    public boolean canMove(double speed) {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:36]: why canMove?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 12]
```diff
@@ -1,4 +1,44 @@
...
+        return true;
+    }
+
+    @Override
+    public void stop() {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:41]: why stop?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 13]
```diff
@@ -1,4 +1,29 @@
...
+    private static Transfer instance;
+
+    private Transfer(WPI_TalonSRX talon) {
+        super("transfer", talon);
+        this.talon = talon;
+        this.entranceSensor = new DigitalInput(RobotMap.DIO.TRANSFER_ENTRANCE_LIGHT_SENSOR);
+        this.exitSensor = new DigitalInput(RobotMap.DIO.TRANSFER_EXIT_LIMIT_SWITCH);
+    }
+
+    public static Transfer getInstance() {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:23]: not the right place for the `getInstance()` function

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 14]
```diff
@@ -1,4 +1,29 @@
...
+        if (instance == null) {
+            instance = new Transfer(new WPI_TalonSRX(RobotMap.CAN.TRANSFER_TALON));
+        }
+        return instance;
+    }
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:28]: add `configureDashboard()`

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 15]
```diff
@@ -1,4 +1,35 @@
...
 
-public class Transfer {
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.spikes2212.command.genericsubsystem.GenericSubsystem;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.RobotMap;
+
+public class Transfer extends MotoredGenericSubsystem {
+    private WPI_TalonSRX talon;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:10]: why do u save this in here? its already in super

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 16]
```diff
@@ -1,4 +1,38 @@
 package frc.robot.Subsystems;
 
-public class Transfer {
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.spikes2212.command.genericsubsystem.GenericSubsystem;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.RobotMap;
+
+public class Transfer extends MotoredGenericSubsystem {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:9]: add docs

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 17]
```diff
@@ -1,4 +1,41 @@
...
 
-public class Transfer {
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.spikes2212.command.genericsubsystem.GenericSubsystem;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.RobotMap;
+
+/**
+ * uses timing straps to transport cargo pieces from the intake to the dispenser.
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:10]: too mechanical

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 18]
```diff
@@ -1,4 +1,41 @@
...
+
+    public static Transfer getInstance() {
+        if (instance == null) {
+            instance = new Transfer(new WPI_TalonSRX(RobotMap.CAN.TRANSFER_TALON));
+        }
+        return instance;
+    }
+
+    private Transfer(WPI_TalonSRX talon) {
+        super("transfer", talon);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:28]: there are 2 talons in this subsystem

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 19]
```diff
@@ -1,4 +1,47 @@
...
+    private final DigitalInput entranceSensor;
+    // a light sensor that sends a signal when a cargo piece exits the system
+    private final DigitalInput exitSensor;
+    private final Limelight limelight;
+
+    private static Transfer instance;
+
+    public static Transfer getInstance() {
+        if (instance == null) {
+            instance = new Transfer(new WPI_TalonSRX(RobotMap.CAN.TRANSFER_TALON_1), new WPI_TalonSRX(RobotMap.CAN.TRANSFER_TALON_2));
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:23]: one of the should be inverted

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 20]
```diff
@@ -1,4 +1,48 @@
...
+        talon2.setInverted(true);
+        this.entranceSensor = new DigitalInput(RobotMap.DIO.TRANSFER_ENTRANCE_LIGHT_SENSOR);
+        this.exitSensor = new DigitalInput(RobotMap.DIO.TRANSFER_EXIT_LIMIT_SWITCH);
+        this.limelight = new Limelight();
+    }
+
+    public Limelight getLimelight() {
+        return limelight;
+    }
+
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:39]: getters for the sensors get functions

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Comment by tal-sitton** [comment]
This subsystem has been updated, please update the code accordingly

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 21]
```diff
@@ -1,4 +1,94 @@
...
+            );
+        }
+        return instance;
+    }
+
+    private Transfer(WPI_TalonSRX entranceTalon, WPI_TalonSRX strapTalon1, WPI_TalonSRX strapTalon2) {
+        this.entranceMotorSensor = new DigitalInput(RobotMap.DIO.TRANSFER_WHEEL_ENTRANCE_LIMIT_SWITCH);
+        this.strapEntranceSensor = new DigitalInput(RobotMap.DIO.TRANSFER_STRAP_LIGHT_SENSOR);
+        this.limelight = new Limelight();
+        this.strapMotors = new MotorControllerGroup(strapTalon1, strapTalon2);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:46]: one should be inverted

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 22]
```diff
@@ -1,4 +1,94 @@
...
+            instance = new Transfer(new WPI_TalonSRX(RobotMap.CAN.TRANSFER_ENTRANCE_MOTOR),
+                    new WPI_TalonSRX(RobotMap.CAN.TRANSFER_STRAP_TALON_1),
+                    new WPI_TalonSRX(RobotMap.CAN.TRANSFER_STRAP_TALON_2)
+            );
+        }
+        return instance;
+    }
+
+    private Transfer(WPI_TalonSRX entranceTalon, WPI_TalonSRX strapTalon1, WPI_TalonSRX strapTalon2) {
+        this.entranceMotorSensor = new DigitalInput(RobotMap.DIO.TRANSFER_WHEEL_ENTRANCE_LIMIT_SWITCH);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:43]: the first lines in the constructor should be connected to the objects that you get in the ()

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 23]
```diff
@@ -1,4 +1,94 @@
...
+        this.strapEntranceSensor = new DigitalInput(RobotMap.DIO.TRANSFER_STRAP_LIGHT_SENSOR);
+        this.limelight = new Limelight();
+        this.strapMotors = new MotorControllerGroup(strapTalon1, strapTalon2);
+        this.entranceTalon = entranceTalon;
+    }
+//    private Transfer(WPI_TalonSRX talon1, WPI_TalonSRX talon2) {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:49]: ?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 24]
```diff
@@ -1,4 +1,94 @@
...
+            entranceTalon.set(entranceMotorSpeed.get());
+        } else {
+            strapMotors.set(strapSpeed.get());
+            entranceTalon.set(entranceMotorSpeed.get());
+        }
+    }
+
+    @Override
+    public boolean canMove(double speed) {
+        return true;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:69]: is it always true?

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 25]
```diff
@@ -1,4 +1,94 @@
...
+    public void stop() {
+        entranceTalon.stopMotor();
+        strapMotors.stopMotor();
+    }
+
+    public Limelight getLimelight() {
+        return limelight;
+    }
+
+    public boolean getEntranceMotorLimitPressed() {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:82]: remove `Pressed`

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 26]
```diff
@@ -1,4 +1,94 @@
...
+    public boolean canMove(double speed) {
+        return true;
+    }
+
+    @Override
+    public void stop() {
+        entranceTalon.stopMotor();
+        strapMotors.stopMotor();
+    }
+
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:77]: add `periodic`

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 27]
```diff
@@ -1,4 +1,60 @@
...
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
+import frc.robot.RobotMap;
+
+import java.util.function.Supplier;
+
+/**
+ * transports cargo from the bottom to the top of the robot
+ */
+public class Transfer extends GenericSubsystem {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Transfer.java:16]: MotoredGericSubsystem

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java [hunk 28]
```diff
@@ -19,10 +20,10 @@
      * This function is run when the robot is first started up and should be used for any
      * initialization code.
      */
+    private Transfer transfer;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Robot.java:23]: add blank line after this

File: src/main/java/frc/robot/RobotMap.java [hunk 29]
```diff
@@ -1,4 +1,12 @@
 package frc.robot;
 
 public class RobotMap {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/RobotMap.java:3]: blank line at the start of the class

File: src/main/java/frc/robot/RobotMap.java [hunk 30]
```diff
@@ -1,4 +1,12 @@
...
+    public interface CAN {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/RobotMap.java:4]: blank line at the start of the interface

File: src/main/java/frc/robot/RobotMap.java [hunk 31]
```diff
@@ -1,4 +1,12 @@
...
+        int TRANSFER_STRAP_TALON_1 = -1;
+        int TRANSFER_STRAP_TALON_2 = -1;
+    }
+
+    public interface DIO {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/RobotMap.java:9]: blank line at the start of the interface

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 32]
```diff
@@ -1,4 +1,41 @@
...
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.util.Limelight;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.RobotMap;
+
+/**
+ * transports cargo from the bottom to the top of the robot
+ */
+public class Transfer extends MotoredGenericSubsystem {
+    // a light sensor that sends a signal while a cargo is held at the bottom of the timing straps
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/Transfer.java:13]: do documentation in javadoc

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 33]
```diff
@@ -1,4 +1,41 @@
...
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.util.Limelight;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.RobotMap;
+
+/**
+ * transports cargo from the bottom to the top of the robot
+ */
+public class Transfer extends MotoredGenericSubsystem {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/Transfer.java:12]: add blank line at the start of the class

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 34]
```diff
@@ -1,4 +1,41 @@
...
 
-public class Transfer {
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.spikes2212.command.genericsubsystem.MotoredGenericSubsystem;
+import com.spikes2212.util.Limelight;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.RobotMap;
+
+/**
+ * transports cargo from the bottom to the top of the robot
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/Transfer.java:10]: Capital letter at the start of the sentence

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 35]
```diff
@@ -1,4 +1,41 @@
...
+    public static Transfer getInstance() {
+        if (instance == null) {
+            instance = new Transfer(new WPI_TalonSRX(RobotMap.CAN.TRANSFER_STRAP_TALON_1),
+                    new WPI_TalonSRX(RobotMap.CAN.TRANSFER_STRAP_TALON_2)
+            );
+        }
+        return instance;
+    }
+
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/Transfer.java:29]: remove blank line

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java (hunk 28 above)
Comment (yutytuty) [reply src/main/java/frc/robot/Robot.java:23]: wrong grammer

File: src/main/java/frc/robot/RobotMap.java (hunk 29 above)
Comment (yutytuty) [reply src/main/java/frc/robot/RobotMap.java:3]: grammer

File: src/main/java/frc/robot/RobotMap.java (hunk 30 above)
Comment (yutytuty) [reply src/main/java/frc/robot/RobotMap.java:4]: ditto

File: src/main/java/frc/robot/RobotMap.java (hunk 31 above)
Comment (yutytuty) [reply src/main/java/frc/robot/RobotMap.java:9]: ditto

File: src/main/java/frc/robot/Subsystems/Transfer.java (hunk 32 above)
Comment (yutytuty) [reply src/main/java/frc/robot/Subsystems/Transfer.java:13]: ditto

File: src/main/java/frc/robot/RobotMap.java (hunk 31 above)
Comment (OfriRosenbaum) [reply src/main/java/frc/robot/RobotMap.java:9]: add please

File: src/main/java/frc/robot/RobotMap.java (hunk 30 above)
Comment (OfriRosenbaum) [reply src/main/java/frc/robot/RobotMap.java:4]: add please

File: src/main/java/frc/robot/RobotMap.java (hunk 29 above)
Comment (OfriRosenbaum) [reply src/main/java/frc/robot/RobotMap.java:3]: add please

File: src/main/java/frc/robot/Robot.java (hunk 28 above)
Comment (OfriRosenbaum) [reply src/main/java/frc/robot/Robot.java:23]: grammar*

File: src/main/java/frc/robot/RobotMap.java (hunk 29 above)
Comment (OfriRosenbaum) [reply src/main/java/frc/robot/RobotMap.java:3]: grammar*

File: src/main/java/frc/robot/RobotMap.java (hunk 30 above)
Comment (OfriRosenbaum) [reply src/main/java/frc/robot/RobotMap.java:4]: > ditto

grammar*

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 36]
```diff
@@ -7,10 +7,13 @@
...
 
 /**
- * transports cargo from the bottom to the top of the robot
+ * Transports cargo from the bottom to the top of the robot
  */
 public class Transfer extends MotoredGenericSubsystem {
-    // a light sensor that sends a signal while a cargo is held at the bottom of the timing straps
+
+    /**
+     * A light sensor that sends a signal while a cargo is held at the bottom of the timing straps
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/Transfer.java:15]: add a period at the end of this line

File: src/main/java/frc/robot/Subsystems/Transfer.java [hunk 37]
```diff
@@ -7,10 +7,13 @@
 import frc.robot.RobotMap;
 
 /**
- * transports cargo from the bottom to the top of the robot
+ * Transports cargo from the bottom to the top of the robot
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/Transfer.java:10]: add a period at the end of this line

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

**Review by OfriRosenbaum — APPROVED** [review]

File: src/main/java/frc/robot/Robot.java [hunk 38]
```diff
@@ -34,10 +36,6 @@ public void robotInit() {
      */
     @Override
     public void robotPeriodic() {
-        // Runs the Scheduler.  This is responsible for polling buttons, adding newly-scheduled
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:37]: why?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Review by tal-sitton — APPROVED** [review]

**Pull request merged**